File: Frontend/src/app/core/passenger.service.ts

import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../environments/environment';

export interface PassengerDto {
  id: number;
  name: string;
  email: string;
  mobile: string;
  address: string;
  dob: string; // ISO (yyyy-MM-dd)
}
export interface UpdatePassengerRequest {
  name: string;
  mobile: string;
  address: string;
}

@Injectable({ providedIn: 'root' })
export class PassengerService {
  private http = inject(HttpClient);
  private base = `${environment.apiUrl}/passenger`; // NOTE: singular!

  getById(id: number) {
    return this.http.get<PassengerDto>(`${this.base}/${id}`);
  }
  update(id: number, req: UpdatePassengerRequest) {
    return this.http.put<PassengerDto>(`${this.base}/${id}`, req);
  }
}

File: Backend/src/main/java/com/ilp/trainticket/service/PassengerService.java

package com.ilp.trainticket.service;

import org.springframework.stereotype.Service;
import com.ilp.trainticket.repository.PassengerRepository;
import com.ilp.trainticket.entity.Passenger;
import com.ilp.trainticket.dto.*;
import com.ilp.trainticket.util.PasswordUtil;
import com.ilp.trainticket.exception.ApiException;

@Service
public class PassengerService {

    private final PassengerRepository repo;

    public PassengerService(PassengerRepository repo) {
        this.repo = repo;
    }

    public Passenger register(RegisterRequest r) {
        if (repo.existsByEmail(r.email)) {
            throw new ApiException("Email already registered");
        }
        Passenger p = new Passenger();
        p.setName(r.name);
        p.setEmail(r.email);
        p.setMobile(r.mobile);
        p.setAddress(r.address);
        p.setDob(r.dob);
        p.setPassword(PasswordUtil.hash(r.password));
        return repo.save(p);
    }

    public Passenger login(LoginRequest r) {
        Passenger p = repo.findAll().stream()
                .filter(x -> x.getEmail().equals(r.email))
                .findFirst()
                .orElseThrow(() -> new ApiException("Invalid email or password"));

        if (!PasswordUtil.match(r.password, p.getPassword())) {
            throw new ApiException("Invalid email or password");
        }
        return p;
    }

    // ===== NEW: Profile read/update =====
    public PassengerDto getById(Long id) {
        Passenger p = repo.findById(id).orElseThrow(() -> new ApiException("Passenger not found"));
        return new PassengerDto(p.getId(), p.getName(), p.getEmail(), p.getMobile(), p.getAddress(), p.getDob());
    }

    public PassengerDto updateById(Long id, UpdatePassengerRequest req) {
        Passenger p = repo.findById(id).orElseThrow(() -> new ApiException("Passenger not found"));
        if (req.name() != null)   p.setName(req.name());
        if (req.mobile() != null) p.setMobile(req.mobile());
        if (req.address() != null)p.setAddress(req.address());
        Passenger saved = repo.save(p);
        return new PassengerDto(saved.getId(), saved.getName(), saved.getEmail(), saved.getMobile(), saved.getAddress(), saved.getDob());
    }
}
